src/storage/mod.rs

```diff
@@ -1107,44 +1108,58 @@ impl<E: Engine> Storage<E> {
-                                stats.data.flow_stats.read_bytes = key_len + value.len();
-                                thread_ctx.collect_read_flow(ctx.get_region_id(), &stats);
-                                thread_ctx.collect_key_reads(CMD, 1);
-                            }
-                            r
-                        })
-                })
-                .then(move |r| {
-                    _timer.observe_duration();
-                    r
-                })
-        });
+        let _timer = SCHED_HISTOGRAM_VEC
+            .with_label_values(&[CMD])
+            .start_coarse_timer();
+
+        let (tx, future) = oneshot::channel();
+        let readpool = self.read_pool.clone();
+        let region_id = ctx.get_region_id();
+
+        let cb = box move |(_, snapshot)| {
+            if let Ok(fut) = readpool.future_execute(priority, move |ctxd| {
+                future::result(snapshot)
+                    .map_err(txn::Error::from)
+                    .map_err(Error::from)
+                    .and_then(move |snapshot: E::Snap| {
+                        let mut thread_ctx = ctxd.current_thread_context_mut();
+                        let _t_process = thread_ctx.start_processing_read_duration_timer(CMD);
+                        let cf = Self::rawkv_cf(&cf)?;
+                        // no scan_count for this kind of op.
+
+                        let key_len = key.len();
+                        snapshot.get_cf(cf, &Key::from_encoded(key))
+                            // map storage::engine::Error -> storage::Error
+                            .map_err(Error::from)
+                            .map(|r| {
+                                if let Some(ref value) = r {
+                                    let mut stats = Statistics::default();
+                                    stats.data.flow_stats.read_keys = 1;
+                                    stats.data.flow_stats.read_bytes = key_len + value.len();
+                                    thread_ctx.collect_read_flow(region_id, &stats);
+                                    thread_ctx.collect_key_reads(CMD, 1);
+                                }
+                                r
+                            })
+                    })
+                    .and_then(move |res| {
+                        _timer.observe_duration();
+                        let r = tx.send(res);
+                        if r.is_err() {
+                            warn!("future callback: Failed to send result to the future rx, discarded.");
+                        }
+                        Ok(())
+                    })
+            }) {
+                fut.forget();
+            }
+        };
 
+        let res = engine.async_snapshot(&ctx, cb);
         future::result(res)
-            .map_err(|_| Error::SchedTooBusy)
+            .map(|_| future.map_err(|e| EngineError::Other(box_err!(e))))
             .flatten()
+            .map_err(|_| Error::SchedTooBusy)
```

Comment:
This can be wrong as the result returned by `async_snapshot` can be various and needs to be handled properly to let client choose properly retry strategy.

Reply:
Yes, this error handling will be changed properly.